### PR TITLE
FEAT: Added load_data to sqlalchemy's backend

### LINKED ISSUE #1979: 'PostgreSQLClient' object has no attribute 'engine'
hey everyone!

`create_table` with postgres backend is raising an error:

```
Traceback (most recent call last):
  File "winop.py", line 68, in <module>
    create_psql_table(psql, df)
  File "winop.py", line 44, in create_psql_table
    psql.create_table('salary', schema=sch, database='ibis_testing')
  File "/mnt/sda1/dev/quansight/ibis-project/ibis/ibis/sql/alchemy.py", line 1022, in wrapped
    result = f(self, *args, **kwargs)
  File "/mnt/sda1/dev/quansight/ibis-project/ibis/ibis/sql/alchemy.py", line 1058, in create_table
    if database is not None and database != self.engine.url.database:
AttributeError: 'PostgreSQLClient' object has no attribute 'engine'

```

the code used is:

```python
# ...

def create_psql_table(psql, df):
    sch = ibis.schema([
        ('first_name', 'string'),
        ('last_name', 'string'),
        ('department_name', 'string'),
        ('salary', 'float64'),
    ])
    psql.create_table('salary', schema=sch, database='ibis_testing')
    psql.load_data('salary', df, database='ibis_testing')

psql = make_psql_connection()
create_psql_table(psql, df)

```

env:

```
# packages in environment at /mnt/sda1/storage/miniconda/envs/ibis-dev:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
apipkg                    1.5                        py_0    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
arrow-cpp                 0.13.0           py37h6b969ab_7    conda-forge
asn1crypto                0.24.0                py37_1003    conda-forge
aspy-refactor-imports     1.1.0                    pypi_0    pypi
aspy.yaml                 1.3.0                      py_0    conda-forge
atomicwrites              1.3.0                      py_0    conda-forge
attrs                     19.1.0                     py_0    conda-forge
bcrypt                    3.1.6            py37h516909a_1    conda-forge
bitarray                  1.0.1            py37h516909a_0    conda-forge
black                     19.3b0                     py_0    conda-forge
blinker                   1.4                        py_1    conda-forge
blosc                     1.17.0               he1b5a44_1    conda-forge
boost-cpp                 1.70.0               h8e57a91_2    conda-forge
brotli                    1.0.7             he1b5a44_1000    conda-forge
bzip2                     1.0.8                h516909a_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
cached-property           1.5.1                      py_0    conda-forge
cachetools                2.1.0                      py_0    conda-forge
cairo                     1.16.0            hfb77d84_1002    conda-forge
certifi                   2019.9.11                py37_0    conda-forge
cffi                      1.12.3           py37h8022711_0    conda-forge
cfgv                      2.0.1                      py_0    conda-forge
cfitsio                   3.470                hb60a0a2_2    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
clickhouse-cityhash       1.0.2.3          py37he1b5a44_0    conda-forge
clickhouse-driver         0.1.1            py37h516909a_0    conda-forge
clickhouse-sqlalchemy     0.0.10                     py_0    conda-forge
cligj                     0.5.0                      py_0    conda-forge
cmake                     3.15.3               hf94ab9c_0    conda-forge
coverage                  4.5.4            py37h516909a_0    conda-forge
cryptography              2.7              py37h72c5cf5_0    conda-forge
curl                      7.65.3               hf8cf82a_0    conda-forge
docopt                    0.6.2                      py_1    conda-forge
double-conversion         3.1.5                he1b5a44_1    conda-forge
editdistance              0.5.3            py37hf484d3e_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
execnet                   1.7.1                      py_0    conda-forge
expat                     2.2.5             he1b5a44_1003    conda-forge
fastavro                  0.22.5           py37h516909a_0    conda-forge
fiona                     1.8.7            py37hdff7cfa_0    conda-forge
flake8                    3.7.8                    py37_1    conda-forge
fontconfig                2.13.1            h86ecdb6_1001    conda-forge
freetype                  2.10.0               he983fc9_1    conda-forge
freexl                    1.0.5             h14c3975_1002    conda-forge
fribidi                   1.0.5             h516909a_1002    conda-forge
gdal                      2.4.2            py37h5f563d9_4    conda-forge
geoalchemy2               0.6.3                      py_0    conda-forge
geopandas                 0.5.1                      py_0    conda-forge
geos                      3.7.2                he1b5a44_2    conda-forge
geotiff                   1.5.1                h560c3f3_2    conda-forge
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
gflags                    2.2.2             he1b5a44_1001    conda-forge
giflib                    5.1.7                h516909a_1    conda-forge
glib                      2.58.3            h6f030ca_1002    conda-forge
glog                      0.4.0                he1b5a44_1    conda-forge
google-api-core           1.14.2                   py37_0    conda-forge
google-auth               1.6.3                      py_0    conda-forge
google-auth-oauthlib      0.4.0                      py_0    conda-forge
google-cloud-bigquery     1.19.0                   py37_0    conda-forge
google-cloud-core         1.0.3                      py_0    conda-forge
google-resumable-media    0.4.1                      py_0    conda-forge
googleapis-common-protos  1.6.0                    py37_0    conda-forge
graphite2                 1.3.13            hf484d3e_1000    conda-forge
graphviz                  2.40.1               h0f2764d_1    conda-forge
harfbuzz                  2.4.0                h9f30f68_3    conda-forge
hdf4                      4.2.13            h9a582f1_1002    conda-forge
hdf5                      1.10.5          nompi_h3c11f04_1103    conda-forge
ibis-framework            1.2.0+53.gae0472ab.dirty           dev_0    <develop>
icu                       64.2                 he1b5a44_1    conda-forge
identify                  1.4.7                      py_0    conda-forge
idna                      2.8                   py37_1000    conda-forge
importlib_metadata        0.23                     py37_0    conda-forge
impyla                    0.15.0                   py37_0    conda-forge
isort                     4.3.21                   py37_0    conda-forge
jinja2                    2.10.1                     py_0    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
json-c                    0.13.1            h14c3975_1001    conda-forge
kealib                    1.4.10            h58c409b_1005    conda-forge
krb5                      1.16.3            h05b26f9_1001    conda-forge
libblas                   3.8.0               12_openblas    conda-forge
libcblas                  3.8.0               12_openblas    conda-forge
libcurl                   7.65.3               hda55be3_0    conda-forge
libdap4                   3.20.4               hd3bb157_0    conda-forge
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libevent                  2.1.10               h72c5cf5_0    conda-forge
libffi                    3.2.1             he1b5a44_1006    conda-forge
libgcc-ng                 9.1.0                hdf63c60_0  
libgdal                   2.4.2                hbef8c27_4    conda-forge
libgfortran-ng            7.3.0                hdf63c60_0  
libgit2                   0.28.3               h241e3f0_0    conda-forge
libiconv                  1.15              h516909a_1005    conda-forge
libkml                    1.3.0             h4fcabce_1010    conda-forge
liblapack                 3.8.0               12_openblas    conda-forge
libnetcdf                 4.6.2             h303dfb8_1003    conda-forge
libopenblas               0.3.7                h6e990d7_1    conda-forge
libpng                    1.6.37               hed695b0_0    conda-forge
libpq                     11.5                 hd9ab2ff_1    conda-forge
libprotobuf               3.8.0                h8b12597_0    conda-forge
libspatialindex           1.9.0                he1b5a44_1    conda-forge
libspatialite             4.3.0a            he1bb1e1_1029    conda-forge
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.0.10            h57b8799_1003    conda-forge
libtool                   2.4.6             h14c3975_1002    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libuv                     1.31.0               h516909a_0    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.9                hee79883_5    conda-forge
lz4                       2.2.1            py37hd79334b_0    conda-forge
lz4-c                     1.8.3             he1b5a44_1001    conda-forge
lzo                       2.10              h14c3975_1000    conda-forge
markupsafe                1.1.1            py37h14c3975_0    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
mock                      3.0.5                    py37_0    conda-forge
more-itertools            7.2.0                      py_0    conda-forge
multipledispatch          0.6.0                    pypi_0    pypi
munch                     2.3.2                      py_0    conda-forge
mypy                      0.720                      py_0    conda-forge
mypy_extensions           0.4.1                 py37_1000    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
nodeenv                   1.3.3                      py_0    conda-forge
numexpr                   2.7.0            py37hb3f55d8_0    conda-forge
numpy                     1.17.2           py37h95a1406_0    conda-forge
oauthlib                  3.0.1                      py_0    conda-forge
openjdk                   8.0.192           h14c3975_1003    conda-forge
openjpeg                  2.3.1                h21c5421_1    conda-forge
openssl                   1.1.1c               h516909a_0    conda-forge
packaging                 19.2                       py_0    conda-forge
pandas                    0.25.1           py37hb3f55d8_0    conda-forge
pango                     1.42.4               ha030887_1    conda-forge
paramiko                  2.6.0                    py37_0    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
pcre                      8.41              hf484d3e_1003    conda-forge
pip                       19.2.3                   py37_0    conda-forge
pixman                    0.38.0            h516909a_1003    conda-forge
pluggy                    0.13.0                   py37_0    conda-forge
plumbum                   1.6.7                      py_0    conda-forge
ply                       3.11                       py_1    conda-forge
poppler                   0.67.0               ha967d66_7    conda-forge
poppler-data              0.4.9                         1    conda-forge
postgresql                11.5                 hc63931a_1    conda-forge
pre_commit                1.18.1                   py37_0    conda-forge
proj4                     6.1.0                he751ad9_2    conda-forge
protobuf                  3.8.0            py37he1b5a44_1    conda-forge
psutil                    5.6.3            py37h516909a_0    conda-forge
psycopg2                  2.8.3            py37h72c5cf5_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
py                        1.8.0                      py_0    conda-forge
py4j                      0.10.7                     py_1    conda-forge
pyarrow                   0.13.0           py37h8b68381_2    conda-forge
pyasn1                    0.4.6                      py_0    conda-forge
pyasn1-modules            0.2.6                      py_0    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pydata-google-auth        0.1.3                      py_0    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pygit2                    0.28.2           py37h516909a_0    conda-forge
pyjwt                     1.7.1                      py_0    conda-forge
pykerberos                1.2.1            py37h14c3975_0  
pymapd                    0.16.0                     py_0    conda-forge
pymysql                   0.9.3                      py_0    conda-forge
pynacl                    1.3.0           py37h14c3975_1000    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyproj                    2.2.1            py37hc44880f_0    conda-forge
pysocks                   1.7.1                    py37_0    conda-forge
pyspark                   2.4.4                      py_0    conda-forge
pytables                  3.5.2            py37h9f153d1_2    conda-forge
pytest                    5.1.2                    py37_0    conda-forge
pytest-cov                2.7.1                      py_0    conda-forge
pytest-forked             1.0.2                      py_0    conda-forge
pytest-xdist              1.29.0                     py_0    conda-forge
python                    3.7.3                h33d41f4_1    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
python-graphviz           0.13                       py_0    conda-forge
python-hdfs               2.5.7                      py_0    conda-forge
pytz                      2019.2                     py_0    conda-forge
pyyaml                    5.1.2            py37h516909a_0    conda-forge
re2                       2019.09.01           he1b5a44_0    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
regex                     2019.08.19       py37h516909a_0    conda-forge
requests                  2.22.0                   py37_1    conda-forge
requests-kerberos         0.12.0                   py37_0  
requests-oauthlib         1.2.0                      py_0    conda-forge
rhash                     1.3.6             h14c3975_1001    conda-forge
rsa                       3.4.2                      py_1    conda-forge
rtree                     0.8.3           py37h666c49c_1002    conda-forge
ruamel.yaml               0.16.5           py37h516909a_1    conda-forge
ruamel.yaml.clib          0.1.2            py37h516909a_0    conda-forge
seed-isort-config         1.9.2                    pypi_0    pypi
setuptools                41.2.0                   py37_0    conda-forge
shapely                   1.6.4           py37hec07ddf_1006    conda-forge
six                       1.12.0                py37_1000    conda-forge
snappy                    1.1.7             he1b5a44_1002    conda-forge
sqlalchemy                1.3.6            py37h516909a_0    conda-forge
sqlite                    3.29.0               hcee41ef_1    conda-forge
thrift                    0.11.0          py37hf484d3e_1001    conda-forge
thrift-cpp                0.12.0            hf3afdfd_1004    conda-forge
thriftpy                  0.3.9           py37h14c3975_1001    conda-forge
thriftpy2                 0.4.6            py37h516909a_0    conda-forge
tk                        8.6.9             hed695b0_1003    conda-forge
toml                      0.10.0                     py_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
typed-ast                 1.4.0            py37h516909a_0    conda-forge
typing_extensions         3.7.4                    py37_0    conda-forge
tzcode                    2019a             h516909a_1002    conda-forge
urllib3                   1.25.6                   py37_0    conda-forge
virtualenv                16.7.5                     py_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
wheel                     0.33.6                   py37_0    conda-forge
xerces-c                  3.2.2             h8412b87_1004    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.8                h516909a_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxpm               3.5.12            h516909a_1002    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-libxt                1.1.5             h516909a_1003    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
yaml                      0.1.7             h14c3975_1001    conda-forge
zipp                      0.6.0                      py_0    conda-forge
zlib                      1.2.11            h516909a_1006    conda-forge
zstd                      1.4.0                h3b9ef0a_0    conda-forge
```

### PR DESCRIPTION
**Pull request by xmnlab**
In this PR:

- Added load_data 
- Allow database parameter for load/create/drop when database parameter is the same than the current database (fixes #1979)
- Added `has_attachment` to `AlchemyClient`, useful for some operations that need to use `schema` when database attachment is used, for example `to_sql`.

### CODE REVIEW

File: ibis/sql/alchemy.py [hunk 1]
```diff
@@ -1128,6 +1136,32 @@ def drop_table(
         except KeyError:  # schemas won't be cached if created with raw_sql
             pass
 
+    def load_data(
+        self, table_name, data, database=None, if_exists='fail', **kwargs
+    ):
+        """
+        Wraps the LOAD DATA DDL statement. Loads data into an OmniSciDB table
```
Comment (ian-r-rose) [inline ibis/sql/alchemy.py]: Looks like an out-of-date comment.

File: ibis/tests/all/test_client.py [hunk 2]
```diff
@@ -6,6 +7,32 @@
...
+    df = pd.DataFrame(
+        {
+            'first_name': ['A', 'B', 'C'],
+            'last_name': ['D', 'E', 'F'],
+            'department_name': ['AA', 'BB', 'CC'],
+            'salary': [100, 200, 300],
+        }
+    )
+    con.drop_table(table_name, force=True)
+    con.create_table(table_name, schema=sch)
```
Comment (ian-r-rose) [inline ibis/tests/all/test_client.py]: Is there any way to create a table from the inferred pandas schema? That is, can one automatically come up with a schema given a dataframe, thereby enforcing that they are consistent?

Comment (xmnlab) [reply ibis/tests/all/test_client.py]: I will take a look about that.

Comment (xmnlab) [reply ibis/tests/all/test_client.py]: current `create_table` works with schema or an expression. so maybe using pandas backend we can pass the expr from that.

**Comment by xmnlab** [comment]
this PR is ready for review. thanks!

File: ibis/tests/all/test_client.py (hunk 2 above)
Comment (xmnlab) [reply ibis/tests/all/test_client.py]: for passing a dataframe to create_table, maybe would be good to add a `data` parameter. I will open an issue for that.

**Comment by xmnlab** [comment]
CI is green now.

File: docs/source/release.rst [hunk 3]
```diff
@@ -15,6 +15,7 @@ Release Notes
 * :feature:`2060` Add initial support for ibis.random function
 * :support:`2107` Added fragment_size to table creation for OmniSciDB
 * :feature:`2117` Add non-nullable info to schema output
+* :feature:`1981` Add load_data to sqlalchemy's backends
```
Comment (jreback) [inline docs/source/release.rst]: actually can you add an entry for the postrgres bug fix, ping on green.

**Comment by xmnlab** [comment]
@jreback CI is green. thanks!

**Comment by xmnlab** [comment]
@jreback a gentle reminder about this PR :)

File: ci/docker-compose.yml [hunk 4]
```diff
@@ -10,7 +10,7 @@ services:
       POSTGRES_PASSWORD: ''
 
   mysql:
-    image: mariadb:10.2
+    image: mariadb:10.4.12
```
Comment (jreback) [inline ci/docker-compose.yml:13]: do we actually need this new? iow are we now not supporting an older version?

generally we should not touch anything in the CI except for a dedicated PR

this way we don’t regress 

also we want to test oldest versions that are supported and not newest

File: ibis/impala/client.py [hunk 5]
```diff
@@ -516,7 +516,7 @@ def insert(
         )
         return self._execute(statement)
 
-    def load_data(self, path, overwrite=False, partition=None):
+    def load_data(self, path, overwrite=False, partition=None, **kwargs):
```
Comment (jreback) [inline ibis/impala/client.py:519]: why do we need kwargs? generally we want to explicitly list all parameters if possible

File: ibis/sql/alchemy.py [hunk 6]
```diff
@@ -1099,6 +1100,10 @@ def begin(self):
 
     @invalidates_reflection_cache
     def create_table(self, name, expr=None, schema=None, database=None):
+        # reset database if it is the same one used by the connection
+        if database == self.database_name:
```
Comment (jreback) [inline ibis/sql/alchemy.py:1104]: why is this needed?

File: ibis/sql/alchemy.py [hunk 7]
```diff
@@ -1172,6 +1181,55 @@ def drop_table(
...
+        ----------
+        table_name : string
+        data: pandas.DataFrame
+        database : string, optional
+        if_exists : string, optional, default 'fail'
+            The values available are: {‘fail’, ‘replace’, ‘append’}
+
+        Raises
+        ------
+        NotImplementedError
```
Comment (jreback) [inline ibis/sql/alchemy.py:1205]: huh?

File: ibis/sql/alchemy.py (hunk 6 above)
Comment (xmnlab) [reply ibis/sql/alchemy.py:1104]: I hit #1979 testing load_data method locally. Just reseting the database variable when it is the same as sef.database_name fixed the issue.

another way to fix that maybe would be replace `self.engine.url.database` by `self.database_name`

File: ibis/sql/alchemy.py (hunk 7 above)
Comment (xmnlab) [reply ibis/sql/alchemy.py:1205]: it is related to this:

```
           raise NotImplementedError(
                'Loading data to a table from a different database is not '
                'yet implemented'
            )
```

probably the original message is much better. I will use that for the documentation.

File: ibis/impala/client.py (hunk 5 above)
Comment (xmnlab) [reply ibis/impala/client.py:519]: as I am adding a test for all backends .. it simplified the test. but I can add specific arguments for each backend.

**Comment by xmnlab** [comment]
this PR is ready again for a new review. thanks!

**Comment by jreback** [comment]
thanks @xmnlab
